engine/particle/src/particle.cpp

```diff
@@ -1016,12 +1030,79 @@ namespace dmParticle
         particle->m_SourceAngularVelocity = emitter_properties[EMITTER_KEY_PARTICLE_ANGULAR_VELOCITY];
     }
 
-    static float unit_tex_coords[] =
+    static inline bool HasLocalPositionAttribute(const ParticleVertexAttributeInfos& attribute_infos)
     {
-            0.0f,1.0f, 0.0f,0.0f, 1.0f,0.0f, 1.0f,1.0f
+        for (int i = 0; i < attribute_infos.m_NumInfos; ++i)
+        {
+            if (attribute_infos.m_Infos[i].m_SemanticType == dmGraphics::VertexAttribute::SEMANTIC_TYPE_POSITION &&
+                attribute_infos.m_Infos[i].m_CoordinateSpace == dmGraphics::COORDINATE_SPACE_LOCAL)
+            {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    static uint8_t* WriteParticleVertex(const ParticleVertexAttributeInfos& attribute_infos, uint8_t* write_ptr, const Vector3& p, const Vector3& p_local, const Vector4& color, float* uv, float page_index)
+    {
+        for (int i = 0; i < attribute_infos.m_NumInfos; ++i)
+        {
+            const ParticleVertexAttributeInfo& info = attribute_infos.m_Infos[i];
+
+            switch(info.m_SemanticType)
+            {
+                case dmGraphics::VertexAttribute::SEMANTIC_TYPE_POSITION:
+                {
+                    if (info.m_CoordinateSpace == dmGraphics::COORDINATE_SPACE_WORLD)
+                    {
+                        memcpy(write_ptr, &p, info.m_ValueByteSize);
+                    }
+                    else if (info.m_CoordinateSpace == dmGraphics::COORDINATE_SPACE_LOCAL)
+                    {
+                        memcpy(write_ptr, &p_local, info.m_ValueByteSize);
+                    }
+                    else assert(0);
+                } break;
+                case dmGraphics::VertexAttribute::SEMANTIC_TYPE_TEXCOORD:
+                {
+                    memcpy(write_ptr, uv, info.m_ValueByteSize);
+                } break;
+                case dmGraphics::VertexAttribute::SEMANTIC_TYPE_COLOR:
+                {
+                    if (info.m_NameHash == VERTEX_STREAM_COLOR)
+                    {
+                        memcpy(write_ptr, &color, info.m_ValueByteSize);
+                    }
+                    else
+                    {
+                        memcpy(write_ptr, info.m_ValuePtr, info.m_ValueByteSize);
+                    }
+                } break;
```

Comment:
As previously mentioned, we need to know if the stream has the name "color" or not, if that is the case we output out the particle color and otherwise we put the binary value from the actual attribute.

Reply:
Isn't this true for the other cases?
If the user set a POSITION or TEXCOORD attribute with name "extra", it will get the predefined position/uv values.
Or isn't is allowed, and should that a build error?

Also, do we want to look at the type?
E.g. if the user really wants to use a single uint32_t instead of float4 for color for bandwidth.

Later, I also feel this "write logic" should be standardized/centralized somehow.

Reply:
Yeah that's true I suppose. The problem I ran into was that the "color" semantic is the only one that has any value in the editor since we decorate the property with a color widget. So the way we had originally setup the semantic types means that I can't decorate an attribute with a color widget, but maybe that's fine and should be a separate "edit type" value or something.. But yes, if we are to conform these vertex writing functions into a shared location I think we should produce the built-in data based on the semantic types.

> Also, do we want to look at the type?
> E.g. if the user really wants to use a single uint32_t instead of float4 for color for bandwidth.

Do you mean with the built-in semantic types? Because the data produced from the content pipeline is narrowed or expanded to the correct type, but the built-in types are not. For the built-in types we need to do data conversion between the floats (it's all float streams I think?) into whatever format. I think we should do it, but I'd rather try to get this PR in first and then take a stab at merging the vertex writing function in a separate PR.